rsconnect/api.py

```diff
@@ -62,7 +88,43 @@ def __init__(self, url: str, remote_name: str):
         self.url = url
         self.remote_name = remote_name
 
-    def handle_bad_response(self, response: HTTPResponse | object) -> None:
+    @overload
+    def handle_bad_response(self, response: HTTPResponse, is_httpresponse: Literal[True]) -> HTTPResponse: ...
+
+    @overload
+    def handle_bad_response(self, response: HTTPResponse | T, is_httpresponse: Literal[False] = False) -> T: ...
```

Comment:
Can you clarify the purpose of the overloads vs. just having the one method with a boolean parameter?

Reply:
The purpose of the overloads is for type narrowing. Keep in mind that this function takes a `response` object and either throws or returns that same object.

- When `is_httpresponse=False` (the default), then the type checker can be certain that, the returned object is _not_ a `HTTPResponse`. (If the object was an `HTTPResponse`, the function throws.) In all cases where this is used, the input type is either a `TypedDict` or a `HTTPResponse`, and so when it returns, we can be certain that the result is the `TypedDict`.
- When `is_httpresponse=True`, then the returned object _could_ be an `HTTPResponse`. I added this because there are some instances where the input is expected to be an `HTTPResponse`. These are cases where the reponse status is 2xx, but the payload is _not_ JSON; instead, it may be the contents of a file.

This stuff is sort of necessary because the `_do_request` method may return either an HTTPResponse or, if the reponse body is JSON, it will convert that to a dict and return that. The mix of possible return values makes some things downstream of that a bit awkward. I think that code would benefit from being restructured in a future PR.